File: android/core-designsystem/src/main/kotlin/dev/omnesis/android/designsystem/components/FileTypeIcon.kt
// SPDX-License-Identifier: AGPL-3.0-or-later
// Copyright (c) 2026 Adrien Conrath

package dev.omnesis.android.designsystem.components

import androidx.compose.foundation.background
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.outlined.Article
import androidx.compose.material.icons.outlined.CalendarMonth
import androidx.compose.material.icons.outlined.Code
import androidx.compose.material.icons.outlined.Description
import androidx.compose.material.icons.outlined.Email
import androidx.compose.material.icons.outlined.FolderZip
import androidx.compose.material.icons.outlined.Image
import androidx.compose.material.icons.outlined.InsertDriveFile
import androidx.compose.material.icons.outlined.Movie
import androidx.compose.material.icons.outlined.MusicNote
import androidx.compose.material.icons.outlined.PictureAsPdf
import androidx.compose.material.icons.outlined.Slideshow
import androidx.compose.material.icons.outlined.TableChart
import androidx.compose.material3.Icon
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.graphics.vector.ImageVector
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.tooling.preview.Preview
import androidx.compose.ui.unit.Dp
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import dev.omnesis.android.designsystem.theme.OmRadius
import dev.omnesis.android.designsystem.theme.OmTheme
import dev.omnesis.android.designsystem.theme.OmnesisTheme

/**
 * File-type kinds mirroring the portal's `packages/gateway/portal/js/lib/file-type-icons.js`
 * and the iOS `FileTypeKind`. We resolve the kind from MIME type first, then filename
 * extension, then fall back to a generic file. Each kind has a recognisable colour (red PDF,
 * blue doc, green sheet, …) and a Material Outlined icon echoing the portal's Lucide glyph.
 */
enum class FileTypeKind {
    PDF, DOC, SHEET, SLIDE, IMAGE, ARCHIVE, CALENDAR, EMAIL, CODE, AUDIO, VIDEO, TEXT, FILE;

    /** Material Outlined icon echoing the portal's Lucide-style glyph for this kind. */
    val icon: ImageVector
        get() = when (this) {
            PDF -> Icons.Outlined.PictureAsPdf
            DOC -> Icons.Outlined.Description
            SHEET -> Icons.Outlined.TableChart
            SLIDE -> Icons.Outlined.Slideshow
            IMAGE -> Icons.Outlined.Image
            ARCHIVE -> Icons.Outlined.FolderZip
            CALENDAR -> Icons.Outlined.CalendarMonth
            EMAIL -> Icons.Outlined.Email
            CODE -> Icons.Outlined.Code
            AUDIO -> Icons.Outlined.MusicNote
            VIDEO -> Icons.Outlined.Movie
            TEXT -> Icons.Outlined.Article
            FILE -> Icons.Outlined.InsertDriveFile
        }

    /**
     * Fixed file-type palette mirrored from the portal — deliberately literal hex, NOT Primer
     * tokens, so the two surfaces share the instant-recognition language.
     */
    val color: Color
        get() = when (this) {
            PDF -> Color(0xFFDC2626)
            DOC -> Color(0xFF2563EB)
            SHEET -> Color(0xFF16A34A)
            SLIDE -> Color(0xFFEA580C)
            IMAGE -> Color(0xFFA855F7)
            ARCHIVE -> Color(0xFFCA8A04)
            CALENDAR -> Color(0xFF0891B2)
            EMAIL -> Color(0xFF64748B)
            CODE -> Color(0xFF64748B)
            AUDIO -> Color(0xFFDB2777)
            VIDEO -> Color(0xFFDB2777)
            TEXT -> Color(0xFF64748B)
            FILE -> Color(0xFF94A3B8)
        }

    /** Long-form kind label used as a fallback when the filename has no extension. */
    val label: String
        get() = when (this) {
            PDF -> "PDF"
            DOC -> "Document"
            SHEET -> "Spreadsheet"
            SLIDE -> "Presentation"
            IMAGE -> "Image"
            ARCHIVE -> "Archive"
            CALENDAR -> "Calendar"
            EMAIL -> "Email"
            CODE -> "Code"
            AUDIO -> "Audio"
            VIDEO -> "Video"
            TEXT -> "Text"
            FILE -> "File"
        }
}

object FileTypeIcons {
    /** MIME → kind. Mirrors `MIME_TO_KIND` in the portal helper. */
    private val MIME_TO_KIND: Map<String, FileTypeKind> = mapOf(
        "application/pdf" to FileTypeKind.PDF,

        "application/msword" to FileTypeKind.DOC,
        "application/vnd.openxmlformats-officedocument.wordprocessingml.document" to FileTypeKind.DOC,
        "application/vnd.oasis.opendocument.text" to FileTypeKind.DOC,
        "application/rtf" to FileTypeKind.DOC,
        "application/vnd.google-apps.document" to FileTypeKind.DOC,

        "application/vnd.ms-excel" to FileTypeKind.SHEET,
        "application/vnd.openxmlformats-officedocument.spreadsheetml.sheet" to FileTypeKind.SHEET,
        "application/vnd.oasis.opendocument.spreadsheet" to FileTypeKind.SHEET,
        "application/vnd.google-apps.spreadsheet" to FileTypeKind.SHEET,
        "text/csv" to FileTypeKind.SHEET,
        "text/tab-separated-values" to FileTypeKind.SHEET,

        "application/vnd.ms-powerpoint" to FileTypeKind.SLIDE,
        "application/vnd.openxmlformats-officedocument.presentationml.presentation" to FileTypeKind.SLIDE,
        "application/vnd.oasis.opendocument.presentation" to FileTypeKind.SLIDE,
        "application/vnd.google-apps.presentation" to FileTypeKind.SLIDE,

        "application/zip" to FileTypeKind.ARCHIVE,
        "application/x-zip-compressed" to FileTypeKind.ARCHIVE,
        "application/x-tar" to FileTypeKind.ARCHIVE,
        "application/x-rar-compressed" to FileTypeKind.ARCHIVE,
        "application/vnd.rar" to FileTypeKind.ARCHIVE,
        "application/x-7z-compressed" to FileTypeKind.ARCHIVE,
        "application/gzip" to FileTypeKind.ARCHIVE,
        "application/x-gzip" to FileTypeKind.ARCHIVE,
        "application/x-bzip2" to FileTypeKind.ARCHIVE,

        "text/calendar" to FileTypeKind.CALENDAR,
        "application/ics" to FileTypeKind.CALENDAR,

        "message/rfc822" to FileTypeKind.EMAIL,
        "application/vnd.ms-outlook" to FileTypeKind.EMAIL,

        "application/json" to FileTypeKind.CODE,
        "application/xml" to FileTypeKind.CODE,
        "text/xml" to FileTypeKind.CODE,
        "application/javascript" to FileTypeKind.CODE,
        "text/javascript" to FileTypeKind.CODE,
        "application/typescript" to FileTypeKind.CODE,
        "application/x-typescript" to FileTypeKind.CODE,
        "text/x-shellscript" to FileTypeKind.CODE,
        "application/x-sh" to FileTypeKind.CODE,
        "application/x-yaml" to FileTypeKind.CODE,
        "text/yaml" to FileTypeKind.CODE,
        "text/x-yaml" to FileTypeKind.CODE,
        "text/html" to FileTypeKind.CODE,
        "text/css" to FileTypeKind.CODE,

        "text/markdown" to FileTypeKind.TEXT,
        "text/plain" to FileTypeKind.TEXT,
    )

    /** Filename extension → kind. Mirrors `EXT_TO_KIND` in the portal helper. */
    private val EXT_TO_KIND: Map<String, FileTypeKind> = mapOf(
        "pdf" to FileTypeKind.PDF,

        "doc" to FileTypeKind.DOC, "docx" to FileTypeKind.DOC, "rtf" to FileTypeKind.DOC,
        "odt" to FileTypeKind.DOC, "pages" to FileTypeKind.DOC,

        "xls" to FileTypeKind.SHEET, "xlsx" to FileTypeKind.SHEET, "csv" to FileTypeKind.SHEET,
        "tsv" to FileTypeKind.SHEET, "ods" to FileTypeKind.SHEET, "numbers" to FileTypeKind.SHEET,

        "ppt" to FileTypeKind.SLIDE, "pptx" to FileTypeKind.SLIDE, "odp" to FileTypeKind.SLIDE,
        "key" to FileTypeKind.SLIDE,

        "png" to FileTypeKind.IMAGE, "jpg" to FileTypeKind.IMAGE, "jpeg" to FileTypeKind.IMAGE,
        "gif" to FileTypeKind.IMAGE, "webp" to FileTypeKind.IMAGE, "bmp" to FileTypeKind.IMAGE,
        "tiff" to FileTypeKind.IMAGE, "tif" to FileTypeKind.IMAGE, "svg" to FileTypeKind.IMAGE,
        "heic" to FileTypeKind.IMAGE, "heif" to FileTypeKind.IMAGE, "ico" to FileTypeKind.IMAGE,
        "avif" to FileTypeKind.IMAGE,

        "zip" to FileTypeKind.ARCHIVE, "tar" to FileTypeKind.ARCHIVE, "gz" to FileTypeKind.ARCHIVE,
        "tgz" to FileTypeKind.ARCHIVE, "rar" to FileTypeKind.ARCHIVE, "7z" to FileTypeKind.ARCHIVE,
        "bz2" to FileTypeKind.ARCHIVE, "xz" to FileTypeKind.ARCHIVE,

        "ics" to FileTypeKind.CALENDAR, "ical" to FileTypeKind.CALENDAR,

        "eml" to FileTypeKind.EMAIL, "msg" to FileTypeKind.EMAIL, "mbox" to FileTypeKind.EMAIL,

        "json" to FileTypeKind.CODE, "xml" to FileTypeKind.CODE, "yaml" to FileTypeKind.CODE,
        "yml" to FileTypeKind.CODE, "js" to FileTypeKind.CODE, "mjs" to FileTypeKind.CODE,
        "cjs" to FileTypeKind.CODE, "ts" to FileTypeKind.CODE, "tsx" to FileTypeKind.CODE,
        "jsx" to FileTypeKind.CODE, "py" to FileTypeKind.CODE, "rb" to FileTypeKind.CODE,
        "go" to FileTypeKind.CODE, "rs" to FileTypeKind.CODE, "java" to FileTypeKind.CODE,
        "c" to FileTypeKind.CODE, "cpp" to FileTypeKind.CODE, "h" to FileTypeKind.CODE,
        "hpp" to FileTypeKind.CODE, "sh" to FileTypeKind.CODE, "zsh" to FileTypeKind.CODE,
        "bash" to FileTypeKind.CODE, "sql" to FileTypeKind.CODE, "toml" to FileTypeKind.CODE,
        "ini" to FileTypeKind.CODE, "conf" to FileTypeKind.CODE, "cfg" to FileTypeKind.CODE,
        "html" to FileTypeKind.CODE, "htm" to FileTypeKind.CODE, "css" to FileTypeKind.CODE,
        "scss" to FileTypeKind.CODE,

        "mp3" to FileTypeKind.AUDIO, "m4a" to FileTypeKind.AUDIO, "wav" to FileTypeKind.AUDIO,
        "flac" to FileTypeKind.AUDIO, "ogg" to FileTypeKind.AUDIO, "aac" to FileTypeKind.AUDIO,
        "opus" to FileTypeKind.AUDIO,

        "mp4" to FileTypeKind.VIDEO, "mov" to FileTypeKind.VIDEO, "webm" to FileTypeKind.VIDEO,
        "mkv" to FileTypeKind.VIDEO, "avi" to FileTypeKind.VIDEO,

        "txt" to FileTypeKind.TEXT, "md" to FileTypeKind.TEXT, "markdown" to FileTypeKind.TEXT,
        "log" to FileTypeKind.TEXT,
    )

    /**
     * Resolve a kind from a MIME type. The `image`, `audio`, `video`, and `text` top-level
     * types fall through to the catch-all branch so an unspecified `image/heic-sequence` still
     * gets the image icon.
     */
    fun kindFromMime(mime: String?): FileTypeKind? {
        if (mime.isNullOrEmpty()) return null
        val lower = mime.trim().lowercase()
        MIME_TO_KIND[lower]?.let { return it }
        if (lower.startsWith("image/")) return FileTypeKind.IMAGE
        if (lower.startsWith("audio/")) return FileTypeKind.AUDIO
        if (lower.startsWith("video/")) return FileTypeKind.VIDEO
        if (lower.startsWith("text/")) return FileTypeKind.TEXT
        return null
    }

    /** Resolve a kind from a filename's extension. */
    fun kindFromFilename(filename: String?): FileTypeKind? {
        if (filename.isNullOrEmpty()) return null
        // Strip query / fragment in case the "filename" is actually a URL.
        val trimmed = filename.split('?', '#').firstOrNull() ?: filename
        val dot = trimmed.lastIndexOf('.')
        if (dot < 0 || dot >= trimmed.length - 1) return null
        val ext = trimmed.substring(dot + 1).lowercase()
        return EXT_TO_KIND[ext]
    }

    /**
     * Resolve a kind from `(mimeType, filename)` — MIME wins, extension is the fallback.
     * Returns null only if neither resolves.
     */
    fun kind(mimeType: String?, filename: String?): FileTypeKind? =
        kindFromMime(mimeType) ?: kindFromFilename(filename)

    /**
     * Short uppercase label. Prefers the filename extension ("DOCX" wins over the kind-level
     * "DOCUMENT") when available — matches the portal so users see the exact format. Falls
     * back to the MIME kind's label, then "FILE".
     */
    fun label(mimeType: String?, filename: String?): String {
        if (filename != null) {
            val trimmed = filename.split('?', '#').firstOrNull() ?: filename
            val dot = trimmed.lastIndexOf('.')
            if (dot >= 0 && dot < trimmed.length - 1) {
                val ext = trimmed.substring(dot + 1).uppercase()
                if (ext.isNotEmpty() && ext.length <= 5) return ext
            }
        }
        kindFromMime(mimeType)?.let { return it.label.uppercase() }
        return "FILE"
    }
}

/**
 * Small coloured icon for an attachment/file. Ported from the iOS `FileTypeIcon`. Default
 * size mirrors the portal's 16dp default; callers pass 14 or 18 where they need denser /
 * heavier presentation. Drawn in the kind's colour inside a `size+2` frame.
 */
@Composable
fun FileTypeIcon(
    mimeType: String?,
    filename: String?,
    modifier: Modifier = Modifier,
    size: Dp = 16.dp,
) {
    val kind = FileTypeIcons.kind(mimeType, filename) ?: FileTypeKind.FILE
    Icon(
        imageVector = kind.icon,
        contentDescription = kind.label,
        tint = kind.color,
        modifier = modifier.size(size + 2.dp),
    )
}

/**
 * The portal's `.file-type-pill` rendered as a Compose chip — coloured icon + short uppercase
 * label ("PDF", "DOCX", "ICS", …). Ported from the iOS `FileTypePill`. Used in search-result
 * rows and document-detail headers wherever the document type is "attachment" / "file".
 */
@Composable
fun FileTypePill(
    mimeType: String?,
    filename: String?,
    modifier: Modifier = Modifier,
    iconSize: Dp = 12.dp,
) {
    val kind = FileTypeIcons.kind(mimeType, filename) ?: FileTypeKind.FILE
    Row(
        horizontalArrangement = Arrangement.spacedBy(4.dp),
        verticalAlignment = Alignment.CenterVertically,
        modifier = modifier
            .clip(RoundedCornerShape(OmRadius.medium))
            .background(OmTheme.colors.bgTertiary)
            .padding(horizontal = 6.dp, vertical = 2.dp),
    ) {
        Icon(
            imageVector = kind.icon,
            contentDescription = null,
            tint = kind.color,
            modifier = Modifier.size(iconSize),
        )
        Text(
            text = FileTypeIcons.label(mimeType, filename),
            style = MaterialTheme.typography.labelSmall.copy(
                fontSize = 10.sp,
                fontWeight = FontWeight.SemiBold,
                letterSpacing = 0.3.sp,
            ),
            color = OmTheme.colors.textSecondary,
        )
    }
}

// MARK: - Previews

private val FILE_SAMPLES: List<Pair<String?, String?>> = listOf(
    "application/pdf" to "report.pdf",
    "application/vnd.openxmlformats-officedocument.wordprocessingml.document" to "notes.docx",
    "application/vnd.openxmlformats-officedocument.spreadsheetml.sheet" to "budget.xlsx",
    "application/vnd.openxmlformats-officedocument.presentationml.presentation" to "deck.pptx",
    "image/png" to "screenshot.png",
    "application/zip" to "archive.zip",
    "text/calendar" to "invite.ics",
    "message/rfc822" to "thread.eml",
    "application/json" to "config.json",
    "audio/mpeg" to "song.mp3",
    "video/mp4" to "clip.mp4",
    "text/plain" to "note.txt",
    null to "unknown.xyz",
)

@Composable
private fun FileTypeGallery() {
    val c = OmTheme.colors
    Column(
        modifier = Modifier.background(c.bgPrimary).padding(16.dp),
        verticalArrangement = Arrangement.spacedBy(8.dp),
    ) {
        FILE_SAMPLES.forEach { (mime, name) ->
            Row(
                horizontalArrangement = Arrangement.spacedBy(12.dp),
                verticalAlignment = Alignment.CenterVertically,
            ) {
                FileTypeIcon(mimeType = mime, filename = name, size = 18.dp)
                Text(name ?: "—", color = c.textPrimary)
                FileTypePill(mimeType = mime, filename = name)
            }
        }
    }
}

@Preview(name = "FileType — dark")
@Composable
private fun FileTypePreviewDark() {
    OmnesisTheme(darkTheme = true) { FileTypeGallery() }
}

@Preview(name = "FileType — light")
@Composable
private fun FileTypePreviewLight() {
    OmnesisTheme(darkTheme = false) { FileTypeGallery() }
}

File: ios/Sources/Omnesis/UI/FileTypeIcon.swift
// SPDX-License-Identifier: AGPL-3.0-or-later
// Copyright (c) 2026 Adrien Conrath

import Foundation
#if canImport(SwiftUI) && canImport(UIKit)
import SwiftUI

/// File-type icons mirroring the portal's `packages/gateway/portal/js/lib/file-type-icons.js`.
/// We resolve the kind from MIME type first, then filename extension, then
/// fall back to a generic file icon. Each kind has a recognisable colour
/// (red PDF, blue doc, green sheet, orange slides, purple image, …) and an
/// SF Symbol that visually echoes the Lucide glyph the portal uses.
///
/// The portal also exposes a short label ("PDF", "DOCX", …) — `FileTypePill`
/// renders the icon + that label as a compact chip, matching the
/// `.file-type-pill` element on the web.
enum FileTypeKind: String, Hashable {
    case pdf, doc, sheet, slide, image, archive, calendar, email, code, audio, video, text, file

    /// SF Symbol that visually echoes the portal's Lucide-style glyph for
    /// this kind.
    var systemImage: String {
        switch self {
        case .pdf: "doc.richtext.fill"
        case .doc: "doc.text.fill"
        case .sheet: "tablecells.fill"
        case .slide: "rectangle.on.rectangle.fill"
        case .image: "photo.fill"
        case .archive: "archivebox.fill"
        case .calendar: "calendar"
        case .email: "envelope.fill"
        case .code: "chevron.left.forwardslash.chevron.right"
        case .audio: "music.note"
        case .video: "play.rectangle.fill"
        case .text: "doc.plaintext.fill"
        case .file: "doc.fill"
        }
    }

    /// Hex colour mirrored from the portal palette so the two surfaces
    /// share the instant-recognition language.
    var color: Color {
        switch self {
        case .pdf: Color(hex: 0xDC2626)
        case .doc: Color(hex: 0x2563EB)
        case .sheet: Color(hex: 0x16A34A)
        case .slide: Color(hex: 0xEA580C)
        case .image: Color(hex: 0xA855F7)
        case .archive: Color(hex: 0xCA8A04)
        case .calendar: Color(hex: 0x0891B2)
        case .email: Color(hex: 0x64748B)
        case .code: Color(hex: 0x64748B)
        case .audio: Color(hex: 0xDB2777)
        case .video: Color(hex: 0xDB2777)
        case .text: Color(hex: 0x64748B)
        case .file: Color(hex: 0x94A3B8)
        }
    }

    /// Long-form kind label used as a fallback when the filename has no
    /// extension. Matches the portal's `ICONS[kind].label`.
    var label: String {
        switch self {
        case .pdf: "PDF"
        case .doc: "Document"
        case .sheet: "Spreadsheet"
        case .slide: "Presentation"
        case .image: "Image"
        case .archive: "Archive"
        case .calendar: "Calendar"
        case .email: "Email"
        case .code: "Code"
        case .audio: "Audio"
        case .video: "Video"
        case .text: "Text"
        case .file: "File"
        }
    }
}

enum FileTypeIcons {
    /// MIME → kind. Mirrors `MIME_TO_KIND` in the portal helper.
    private static let mimeToKind: [String: FileTypeKind] = [
        "application/pdf": .pdf,

        "application/msword": .doc,
        "application/vnd.openxmlformats-officedocument.wordprocessingml.document": .doc,
        "application/vnd.oasis.opendocument.text": .doc,
        "application/rtf": .doc,
        "application/vnd.google-apps.document": .doc,

        "application/vnd.ms-excel": .sheet,
        "application/vnd.openxmlformats-officedocument.spreadsheetml.sheet": .sheet,
        "application/vnd.oasis.opendocument.spreadsheet": .sheet,
        "application/vnd.google-apps.spreadsheet": .sheet,
        "text/csv": .sheet,
        "text/tab-separated-values": .sheet,

        "application/vnd.ms-powerpoint": .slide,
        "application/vnd.openxmlformats-officedocument.presentationml.presentation": .slide,
        "application/vnd.oasis.opendocument.presentation": .slide,
        "application/vnd.google-apps.presentation": .slide,

        "application/zip": .archive,
        "application/x-zip-compressed": .archive,
        "application/x-tar": .archive,
        "application/x-rar-compressed": .archive,
        "application/vnd.rar": .archive,
        "application/x-7z-compressed": .archive,
        "application/gzip": .archive,
        "application/x-gzip": .archive,
        "application/x-bzip2": .archive,

        "text/calendar": .calendar,
        "application/ics": .calendar,

        "message/rfc822": .email,
        "application/vnd.ms-outlook": .email,

        "application/json": .code,
        "application/xml": .code,
        "text/xml": .code,
        "application/javascript": .code,
        "text/javascript": .code,
        "application/typescript": .code,
        "application/x-typescript": .code,
        "text/x-shellscript": .code,
        "application/x-sh": .code,
        "application/x-yaml": .code,
        "text/yaml": .code,
        "text/x-yaml": .code,
        "text/html": .code,
        "text/css": .code,

        "text/markdown": .text,
        "text/plain": .text,
    ]

    /// Filename extension → kind. Mirrors `EXT_TO_KIND` in the portal helper.
    private static let extToKind: [String: FileTypeKind] = [
        "pdf": .pdf,

        "doc": .doc, "docx": .doc, "rtf": .doc, "odt": .doc, "pages": .doc,

        "xls": .sheet, "xlsx": .sheet, "csv": .sheet, "tsv": .sheet,
        "ods": .sheet, "numbers": .sheet,

        "ppt": .slide, "pptx": .slide, "odp": .slide, "key": .slide,

        "png": .image, "jpg": .image, "jpeg": .image, "gif": .image,
        "webp": .image, "bmp": .image, "tiff": .image, "tif": .image,
        "svg": .image, "heic": .image, "heif": .image, "ico": .image, "avif": .image,

        "zip": .archive, "tar": .archive, "gz": .archive, "tgz": .archive,
        "rar": .archive, "7z": .archive, "bz2": .archive, "xz": .archive,

        "ics": .calendar, "ical": .calendar,

        "eml": .email, "msg": .email, "mbox": .email,

        "json": .code, "xml": .code, "yaml": .code, "yml": .code,
        "js": .code, "mjs": .code, "cjs": .code,
        "ts": .code, "tsx": .code, "jsx": .code,
        "py": .code, "rb": .code, "go": .code, "rs": .code, "java": .code,
        "c": .code, "cpp": .code, "h": .code, "hpp": .code,
        "sh": .code, "zsh": .code, "bash": .code,
        "sql": .code, "toml": .code, "ini": .code, "conf": .code, "cfg": .code,
        "html": .code, "htm": .code, "css": .code, "scss": .code,

        "mp3": .audio, "m4a": .audio, "wav": .audio, "flac": .audio,
        "ogg": .audio, "aac": .audio, "opus": .audio,

        "mp4": .video, "mov": .video, "webm": .video, "mkv": .video, "avi": .video,

        "txt": .text, "md": .text, "markdown": .text, "log": .text,
    ]

    /// Resolve a kind from a MIME type. `image/*`, `audio/*`, `video/*`,
    /// and `text/*` fall through to the catch-all branch so an
    /// unspecified `image/heic-sequence` still gets the image icon.
    static func kind(fromMime mime: String?) -> FileTypeKind? {
        guard let mime, !mime.isEmpty else { return nil }
        let lower = mime.trimmingCharacters(in: .whitespacesAndNewlines).lowercased()
        if let direct = mimeToKind[lower] { return direct }
        if lower.hasPrefix("image/") { return .image }
        if lower.hasPrefix("audio/") { return .audio }
        if lower.hasPrefix("video/") { return .video }
        if lower.hasPrefix("text/") { return .text }
        return nil
    }

    /// Resolve a kind from a filename's extension.
    static func kind(fromFilename filename: String?) -> FileTypeKind? {
        guard let filename, !filename.isEmpty else { return nil }
        // Strip query / fragment in case the "filename" is actually a URL.
        let trimmed = filename.split(whereSeparator: { $0 == "?" || $0 == "#" }).first.map(String.init) ?? filename
        guard let dot = trimmed.lastIndex(of: "."), dot < trimmed.index(before: trimmed.endIndex) else {
            return nil
        }
        let ext = trimmed[trimmed.index(after: dot)...].lowercased()
        return extToKind[String(ext)]
    }

    /// Resolve a kind from `(mimeType, filename)` — MIME wins, extension
    /// is the fallback. Returns nil only if neither resolves.
    static func kind(mimeType: String?, filename: String?) -> FileTypeKind? {
        kind(fromMime: mimeType) ?? kind(fromFilename: filename)
    }

    /// Short uppercase label. Prefers the filename extension ("DOCX" wins
    /// over the kind-level "DOCUMENT") when available — matches the
    /// portal so users see the exact format. Falls back to the MIME
    /// kind's label, then "FILE".
    static func label(mimeType: String?, filename: String?) -> String {
        if let filename {
            let trimmed = filename.split(whereSeparator: { $0 == "?" || $0 == "#" }).first.map(String.init) ?? filename
            if let dot = trimmed.lastIndex(of: "."), dot < trimmed.index(before: trimmed.endIndex) {
                let ext = trimmed[trimmed.index(after: dot)...].uppercased()
                if !ext.isEmpty, ext.count <= 5 { return String(ext) }
            }
        }
        if let k = kind(fromMime: mimeType) { return k.label.uppercased() }
        return "FILE"
    }
}

/// Small coloured SF Symbol matching the portal's `FileTypeIcon` glyph.
/// Default size mirrors the portal's 16pt default; callers pass 14 or
/// 18 where they need denser / heavier presentation.
@available(iOS 17.0, *)
struct FileTypeIcon: View {
    let mimeType: String?
    let filename: String?
    var size: CGFloat = 16

    private var resolved: FileTypeKind {
        FileTypeIcons.kind(mimeType: mimeType, filename: filename) ?? .file
    }

    var body: some View {
        Image(systemName: resolved.systemImage)
            .font(.system(size: size))
            .foregroundStyle(resolved.color)
            .frame(width: size + 2, height: size + 2)
            .accessibilityLabel(resolved.label)
    }
}

/// The portal's `.file-type-pill` rendered as a SwiftUI chip — coloured
/// icon + short uppercase label ("PDF", "DOCX", "ICS", …). Used in
/// search-result rows and document detail headers wherever the
/// document type is "attachment" / "file".
@available(iOS 17.0, *)
struct FileTypePill: View {
    let mimeType: String?
    let filename: String?
    var iconSize: CGFloat = 12

    private var kind: FileTypeKind {
        FileTypeIcons.kind(mimeType: mimeType, filename: filename) ?? .file
    }

    var body: some View {
        HStack(spacing: 4) {
            Image(systemName: kind.systemImage)
                .font(.system(size: iconSize))
                .foregroundStyle(kind.color)
            Text(FileTypeIcons.label(mimeType: mimeType, filename: filename))
                .font(.system(size: 10, weight: .semibold))
                .tracking(0.3)
                .foregroundStyle(Theme.textSecondary)
        }
        .padding(.horizontal, 6)
        .padding(.vertical, 2)
        .background(Theme.bgTertiary)
        .clipShape(RoundedRectangle(cornerRadius: Theme.Radius.medium))
        .accessibilityElement(children: .combine)
    }
}

#if DEBUG
@available(iOS 17.0, *)
#Preview("File-type icons + pills") {
    let samples: [(mime: String?, filename: String?)] = [
        ("application/pdf", "report.pdf"),
        ("application/vnd.openxmlformats-officedocument.wordprocessingml.document", "notes.docx"),
        ("application/vnd.openxmlformats-officedocument.spreadsheetml.sheet", "budget.xlsx"),
        ("application/vnd.openxmlformats-officedocument.presentationml.presentation", "deck.pptx"),
        ("image/png", "screenshot.png"),
        ("application/zip", "archive.zip"),
        ("text/calendar", "invite.ics"),
        ("message/rfc822", "thread.eml"),
        ("application/json", "config.json"),
        ("audio/mpeg", "song.mp3"),
        ("video/mp4", "clip.mp4"),
        ("text/plain", "note.txt"),
        (nil, "unknown.xyz"),
    ]
    return ScrollView {
        VStack(alignment: .leading, spacing: 8) {
            ForEach(Array(samples.enumerated()), id: \.offset) { _, sample in
                HStack(spacing: 12) {
                    FileTypeIcon(mimeType: sample.mime, filename: sample.filename, size: 18)
                    Text(sample.filename ?? "—")
                        .font(.system(size: 13))
                        .foregroundStyle(Theme.textPrimary)
                    Spacer()
                    FileTypePill(mimeType: sample.mime, filename: sample.filename)
                }
                .padding(.horizontal, Theme.Spacing.md)
                .padding(.vertical, 4)
            }
        }
        .padding(.vertical, Theme.Spacing.lg)
    }
    .background(Theme.bgPrimary)
    .preferredColorScheme(.dark)
}
#endif
#endif
